### ISSUE TITLE
NotImplementedError intermittently.

### ISSUE THREAD

**Issue by pkudigra** [grimoirelab/perceval#114]
I have a shellscript that activates perceval and then obtains commits for a number of projects and stores it in a file for further processing. During the running of the script I get a 'NotImplementedError' for some of the projects though others work fine. Furthermore, this behavior seems inconsistent because projects that I could obtain commits for in the previous run sometimes give this error when I run the script again.

**Comment by sduenas** [comment]
May you post here that script? I can't determine what it is going on without more information. I can figure you are using the cache mode for any reason which in git it is not supported but I have to get more info to know where the problem is.

**Comment by pkudigra** [comment]
I am not using docker but the python virtualenv (I followed the instructions here - https://www.gitbook.com/book/jgbarah/grimoirelab-training/details). This is the script I am using and it's pretty basic - 

`#!/bin/bash

if [[ $# -ne 2 ]]; then
    echo "Wrong number of input arguments"
    exit 1
fi
input_file=$1
lang=$2
IFS=","
mkdir $lang
mkdir ~/Project/$lang/logs
echo "Making directory $lang and logs"
source ~/Project/venv/perceval/bin/activate
while read ID url name
do
    giturl=`echo $url | sed -e 's/api.//' | sed -e 's/repos\///' | sed -e 's/$/.git/'`
    echo "Retrieving commits for $giturl"
    log_name=`echo $name | sed -e 's/$/_perc.log/'`
    perceval git $giturl > ~/Project/$lang/logs/$log_name    
done < $input_file
`

**Comment by sanacl** [comment]
hi @pkudigra ! thanks for your bug report.

I've just reproduced the error, have a look at the execution below. The bad new is it does not happen always.

- I've installed in my debian 8.x system with "python setup.py install"
- version 877be8336becb13d7fccda72d9171578bb49130b

```
luis@hogaza ~/repos/perceval ±master » perceval phabricator -h                          141 ↵
[2017-01-30 14:27:25,505] - Sir Perceval is on his quest.
Traceback (most recent call last):
  File "/usr/local/bin/perceval", line 4, in <module>
    __import__('pkg_resources').run_script('perceval==0.6.0.dev1', 'perceval')
  File "/usr/local/lib/python3.5/dist-packages/pkg_resources/__init__.py", line 743, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/local/lib/python3.5/dist-packages/pkg_resources/__init__.py", line 1498, in run_script
    exec(code, namespace, namespace)
  File "/usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/EGG-INFO/scripts/perceval", line 188, in <module>
    main()
  File "/usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/EGG-INFO/scripts/perceval", line 105, in main
    cmd = klass(*args.backend_args)
  File "/usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/perceval/backends/phabricator.py", line 443, in __init__
    super().__init__(*args)
  File "/usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/perceval/backend.py", line 274, in __init__
    parser = self.setup_cmd_parser()
  File "/usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/perceval/backend.py", line 343, in setup_cmd_parser
    raise NotImplementedError
NotImplementedError

luis@hogaza ~/repos/perceval ±master » perceval phabricator -h                            1 ↵
[2017-01-30 14:27:31,015] - Sir Perceval is on his quest.
usage: perceval [-h] [--tag TAG] [--from-date FROM_DATE] [-t API_TOKEN]
                [--cache-path CACHE_PATH] [--clean-cache] [--no-cache]
                [--fetch-cache] [-o OUTFILE]
                url

positional arguments:
  url                   URL of the Phabricator server

optional arguments:
  -h, --help            show this help message and exit

general arguments:
  --tag TAG             tag the items generated during the fetching process
  --from-date FROM_DATE
                        fetch items updated since this date

authentication arguments:
  -t API_TOKEN, --api-token API_TOKEN
                        backend authentication token / API key

cache arguments:
  --cache-path CACHE_PATH
                        directory path to the cache
  --clean-cache         clean the cache before the fetching process
  --no-cache            do not store data in the cache
  --fetch-cache         fetch data from the cache

output arguments:
  -o OUTFILE, --output OUTFILE
                        output file

```

**Comment by sduenas** [comment]
As I can see, perceval is accessing an old version of the backends. Backends are now installed under `perceval/backends/core` directory but it is taking them from `perceval/backends/`.

**Comment by sanacl** [comment]
@albertinisg found an installation error:
- I've deleted the python package with pip and check the dir is correctly removed
- installed again using setup.py
.. and this is the result:

```
luis@hogaza:~$ ls /usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/perceval/backends/
bugzilla.py      gerrit.py    jenkins.py    meetup.py       remo2.py
bugzillarest.py  github.py    jira.py       phabricator.py  remo.py
confluence.py    git.py       kitsune.py    pipermail.py    stackexchange.py
core             gmane.py     mbox.py       __pycache__     supybot.py
discourse.py     __init__.py  mediawiki.py  redmine.py      telegram.py
luis@hogaza:~$ ls /usr/local/lib/python3.5/dist-packages/perceval-0.6.0.dev1-py3.5.egg/perceval/backends/core/
askbot.py        discourse.py  gmane.py     mbox.py         pipermail.py  stackexchange.py
bugzilla.py      gerrit.py     __init__.py  mediawiki.py    __pycache__   supybot.py
bugzillarest.py  github.py     jenkins.py   meetup.py       redmine.py    telegram.py
confluence.py    git.py        jira.py      phabricator.py  rss.py
```

**Comment by sduenas** [comment]
@lcanas Looks like you have some deprecated files inside `dist` directory on your copy of perceval repository. You can clean this directory either removing it or running `setup.py clean --all`.

**Comment by sduenas** [comment]
@pkudigra are you using Perceval's pip package? Looks like this package is broken because it installs deprecated files.

**Comment by pkudigra** [comment]
Thanks for your help and yes, I installed perceval using pip. What is the recommended method of installation?

**Comment by sduenas** [comment]
@pkudigra as long as the package is broken I suggest to install cloning the repository and running  `setup.py`. You have more info [here](https://github.com/grimoirelab/perceval/#source-code).

**Comment by jgbarah** [comment]
I've created a new perceval pip package that should fix the issue. Please, in the same virtual environment (or in one afresh), run:

```
pip install perceval==0.6.0.dev1
```

And let us know if that fixes the issue.

Since this package is still labeled as a dev version, it is not installed by pip by default, that's why you need the `==0.6.0.dev1` thing. It it fixes this issue, we will upgrade it to a stable release asap.

**Comment by pkudigra** [comment]
Yes, the script works perfectly with the new package and doesn't generate the error. Thanks!

**Comment by jgbarah** [comment]
Great! I just released a package for 0.6.0, anyway.
